### PR TITLE
feat(ego): scope-stamp architecture — LLM-judged, structure-enforced develop boundary

### PR DESCRIPTION
**Pull request by WingedGuardian**
Replaces the regex-marker develop-gate (merged in #1316) with a scope-stamp architecture: the operate-vs-develop boundary is **judged by the LLM** (a structured `scope` field) and **enforced by code at chokepoints**. This closes the class of defect that ran #1316's review to the 3-round escalation cap.

## Why the marker gate was the wrong shape

Every round-2/round-3 Codex finding on #1316 was one root cause: **semantic classification done with string matching, enforced at a single pipeline stage.** A regex blocklist over English phrasings is never complete (each round found another verb — PR-review, then refactor, then "implement", …), and draft-time gating missed the reconcile `revise` writer entirely (reconcile runs *before* the realist, so a live-mode revise could mutate an operate proposal into develop work that never got scope-judged). You can't patch a wrong shape into a right one.

## The architecture (LLM judges, structure enforces)

**Judge (LLM, per proposal):**
- The **realist** emits `scope: operate|develop` on every genesis-ego proposal (rubric = the existing operate-vs-develop rule 8). The `autonomy.classify_domain` SELF_MODIFY/AUTONOMOUS_BUILD fast-path **overrides** the LLM (a `code_change` action-type is develop regardless of phrasing).
- The **reconcile** `revise` verdict emits scope for the *sharpened* content (reconcile authors it pre-realist, so it must scope its own output).

**Enforce (deterministic, at every chokepoint a proposal crosses):**
- **create** (`create_batch`): an unstamped genesis-ego draft is **dropped** (fail-closed); `develop` + self-dev-disabled is **created-then-tabled** (recoverable, journal-resolved, deduped vs tabled twins); `operate` is persisted with `scope`/`scope_revision`.
- **revise** (`_reconcile_revise`): a sharpen that is unscoped, or develop-while-disabled, is **not applied** — the draft is kept as a survivor so the realist scopes it through the normal path. An operate item can never be silently mutated into develop. (Closes the round-3 revise-bypass.)
- **dispatch-claim** (`claim_proposal_for_dispatch(allow_develop=…)`): a `develop` row is **unclaimable** on BOTH dispatch paths (execution-brief and approved-sweep) while self-dev is disabled — the atomic last line. NULL-scope rows (user-ego / legacy) are never blocked here.

**Fail-closed, uniformly:** realist error, parse failure, and a missing field all reduce to "no stamp" → the same single create-time drop. Nothing unjudged ever reaches the board; the ego re-derives next cycle (blind-drafting re-proposes anything real), so nothing is lost. The flag (`genesis_self_development_enabled`) flips enforcement off to unlock self-development — and because the realist *stamps* rather than *rejects*, flipping it actually lets develop proposals flow through normal approval (the marker gate could not do this coherently).

**Deleted:** `_DEVELOP_MARKER_PATTERNS`, `_flag_develop_scope`, and the replay harness. One judge, one enforcement.

## Schema
Migration **0078**: `ego_proposals.scope` / `scope_revision`, `ego_proposal_revisions.scope` (dark-column pattern — DDL + base-path ALTER + rebuild-CREATE + numbered migration, all idempotent). Backfill grandfathers genesis-ego pending rows that already passed the realist to `operate` (this install has zero such rows; written for the general case). No `test_schema.py` change (column-level, not a new table).

Also carried: the deliverable floor is now a shared `ensure_deliverable_spec()` applied on BOTH proposal writers (create + revise), closing round-3 P2 (a live-revise could previously strip a valid `expected_outputs` to an unverifiable one).

## Verification
- New `tests/ego/test_scope_stamp.py` (25) + `tests/test_db/test_ego_proposal_scope_migration.py` (4): realist/reconcile scope contract + parse; SELF_MODIFY fast-path override; realist-error → unstamped → dropped; create matrix (drop/table/persist/flag-on/user-ego-untouched); claim guard on develop/operate/null × enabled/disabled; reconcile revise scope gate (operate applies + re-scopes; missing → survivor; develop-disabled → not mutated); prior-scope audit snapshot; global cap counts operate but not develop/unstamped incoming; migration backfill scoping + idempotence.
- `tests/ego/test_reconcile_apply.py` revise verdicts updated to the new scope contract.
- Full `tests/ego/ tests/test_ego/ tests/test_db/` green.
- **Post-merge live-LLM E2E** (per the live-E2E rule for LLM-judgment features): a real genesis-ego cycle with a develop-bait and a symptom-bait directive, asserting the develop draft lands tabled `scope='develop'` (or realist-rejected) and the symptom draft lands pending `scope='operate'`.

**Deploy path:** runtime code + additive idempotent migration 0078 → `git pull` + server restart. Prompt changes take effect next ego cycle. No host paths.

Supersedes the round-3 findings on #1316 (they were the symptom that motivated this redesign).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by WingedGuardian** [comment]
## Code Review — PR #1321

**Structural review:** 1 WARNING, 0 ERRORs  
**PII/Secrets/Wording scan:** CLEAN

---

### WARNING — `_required_outputs_block` lost its unit tests with no replacement

**File:** `src/genesis/ego/session.py` (unchanged) / `tests/ego/test_develop_gate.py` (deleted)

The deleted `test_develop_gate.py` contained `TestRequiredOutputsBlock` — three tests for the `_required_outputs_block` renderer that is still present and active in `session.py`. The replacement `test_scope_stamp.py` covers `ensure_deliverable_spec` and the create/revise/dispatch-claim chokepoints, but has no equivalent coverage for `_required_outputs_block`.

The most fragile case that was deleted:

```python
def test_malformed_shapes_never_raise(self):
    # non-string required_strings must not blow up ', '.join(...)
    block = _required_outputs_block(json.dumps({"files": ["a.md"], "required_strings": 123}))
    assert "a.md" in block  # renders files, drops the bad required_strings
```

The `', '.join(required_strings)` guard on a non-list/non-str input is a real `TypeError` risk if a malformed LLM response ever reaches that path. The other deleted cases (render + empty-input) are lower risk but were explicit coverage of the contract.

Suggest moving the three `TestRequiredOutputsBlock` tests into `test_scope_stamp.py` or a dedicated `test_required_outputs_block.py` so the renderer has pinned edge-case coverage again.

---

### Structural: PASS (with above WARNING)

No pattern violations found across all changed files. All checked items:

- No `nohup` / systemd bypass
- No direct `task_states` inserts
- No `/tmp/` usage
- No bare `except:` clauses — all handlers use `except Exception:` or specific types
- No mutable default arguments
- `scope_guard` f-string in `claim_proposal_for_dispatch` (`ego.py`) is safe — always one of two compile-time string literals from a Python bool, never user-controlled
- Migration 0078: raw SQL is in the migration module (explicitly excepted) and in CRUD (`ego.py`) where it belongs ✓
- Lazy imports inside methods (`classify_domain`, `OutreachCategory`, `load_ego_config`) are intentional — each avoids an import cycle at construction time ✓
- `feat(ego):` conventional commit prefix ✓

### PII/Secrets/Wording: CLEAN

No secrets, private IP ranges, SSH keys, or private hostnames in added lines. `~/.genesis/output/ego-reports/{id}.md` paths use tilde expansion, not hardcoded usernames. PR title and body describe a clean architectural redesign — no security incident language, no credential cleanup references.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/genesis/ego/session.py [hunk 1]
```diff
@@ -1921,6 +1957,33 @@ async def _filter_proposals(
...
+                if self._source_tag == "genesis_ego_cycle":
+                    _scope = verdict.get("scope")
+                    try:
+                        from genesis.autonomy.classification import classify_domain
+                        from genesis.autonomy.types import ActionDomain
+
+                        _dom = classify_domain(
+                            prop.get("action_type", ""),
+                            prop.get("execution_plan") or "",
+                        )
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/ego/session.py:2007]: Fixed in 6e7fe6f3 — _build_realist_prompt now renders the FULL content + execution_plan for genesis-ego proposals (was a 300-char clip with no plan), so the scope judge sees the complete proposal. The SELF_MODIFY fast-path already covered src/ execution_plan refs; this closes free-text/long-content develop intent too. Test: test_genesis_realist_sees_full_content_and_plan.

File: src/genesis/ego/session.py [hunk 2]
```diff
@@ -3309,6 +3284,12 @@ def _due(b: dict) -> str:
 - "new": nothing above covers it — genuinely novel work.
 - "reaffirm": a pending board item ALREADY covers it — put the board id in target_id.
 - "revise": a board item covers it but this draft sharpens/updates it — target_id = board id.
+  For a genesis-ego revise, ALSO emit "scope" ("operate" | "develop") judging the
+  SHARPENED content: OPERATE = diagnose/pull a reversible lever/dispatch a
+  specific-defect investigation whose deliverable is escalated findings; DEVELOP =
+  write/refactor code, change a schema, edit an install script, alter a config
+  *value*, or dev-artifact work even read-only (PR review, tracing source to scope
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/ego/session.py:3342]: Fixed in 6e7fe6f3 — _build_reconcile_prompt renders full content + execution_plan for genesis-ego drafts (was 200-char clip), so the revise scope judgment sees the complete sharpened proposal. Test: test_genesis_reconcile_sees_full_content_and_plan.

File: src/genesis/ego/session.py [hunk 3]
```diff
@@ -2215,12 +2176,20 @@ async def _process_proposals(
...
+                # and all user-ego drafts count. (A genesis-ego develop draft
+                # under a self-development-enabled install is a rare future case
+                # and merely undercounts — erring toward LESS eviction, safe.)
                 _incoming = sum(
-                    1 for p in proposals if not p.get("_develop_scope")
+                    1 for p in proposals
+                    if not (
+                        self._source_tag == "genesis_ego_cycle"
+                        and p.get("_realist_scope") != "operate"
+                    )
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/ego/session.py]: Fixed in 6e7fe6f3 — the cap _incoming predicate now counts a develop draft when self-development is ENABLED (it stays pending, so it's subject to the cap) and excludes it only when disabled (tabled at create); unstamped never counts (dropped). Test: test_develop_incoming_counts_when_enabled.

**Comment by WingedGuardian** [comment]
@codex review — the 3 round-1 findings are fixed in 6e7fe6f3: both P1s (realist + reconcile now feed the scope judge full content + execution_plan for genesis-ego) and the P2 (cap counts develop under flag-on). Please re-review.

File: src/genesis/ego/session.py [hunk 4]
```diff
@@ -1778,18 +1807,25 @@ async def _reconcile_revise(
...
             revised_by=self._source_tag,
             reason=(reason or "reconcile: sharpened from a fresh blind draft"),
             # A revise IS a premise re-validation — advance the cadence clock.
             revalidate_at=next_revalidate_at(
                 board_row.get("urgency", "normal"),
             ),
+            # Re-scope the sharpened content, pinned to the new revision (the
+            # genesis-ego fail-closed checks above guarantee a valid scope here).
+            scope=scope if _is_genesis else None,
+            scope_revision=(expected + 1) if (_is_genesis and scope) else None,
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/ego/session.py:1860]: Fixed in 7c7b07f1 — _reconcile_revise now re-derives scope against the EFFECTIVE (post-COALESCE) execution_plan: the classify_domain SELF_MODIFY/AUTONOMOUS_BUILD fast-path on that plan forces develop, and a board item already stamped develop is never silently downgraded to operate when its plan is retained (draft supplies no replacement). So a retained develop plan can no longer ride under an operate stamp. Test: test_retained_develop_plan_blocks_operate_downgrade.

**Comment by WingedGuardian** [comment]
@codex review — round-2 P1 (scope the merged revision) fixed in 7c7b07f1. Please re-review.
